src/lib.rs

```diff
@@ -173,10 +176,52 @@ impl<T> CircularQueue<T> {
         } else {
             self.data[self.insertion_index] = x;
         }
-
+        if self.reverse_idx > 0 {
+            self.reverse_idx -= 1;
+        }
         self.insertion_index = (self.insertion_index + 1) % self.capacity();
     }
 
+    /// Pops the elements from the queue(LIFO).
+    ///
+    /// Pops at max self.capacity number of items.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use circular_queue::CircularQueue;
+    ///
+    /// let mut queue = CircularQueue::with_capacity(3);
+    /// queue.push(1);
+    /// queue.push(2);
+    /// queue.push(3);
+    /// queue.push(4);
+    ///
+    /// assert_eq!(queue.len(), 3);
+    ///
+    /// assert_eq!(queue.pop(), Some(4));
+    /// assert_eq!(queue.pop(), Some(3));
+    /// assert_eq!(queue.pop(), Some(2));
+    /// assert_eq!(queue.pop(), None);
+    /// 
+    /// ```
+    pub fn pop(&mut self) -> Option<T> {
+        if self.data.len() == 0 {
+            None
+        } else if self.data.len() < self.capacity() {
+            self.data.pop()
+        } else if self.reverse_idx == self.capacity() {
+            None
+        } else {
+            self.reverse_idx += 1;
+            self.insertion_index += self.capacity - 1;
+            self.insertion_index %= self.capacity;
+            unsafe {
+                Some(ptr::read(self.data.get_unchecked(self.insertion_index)))
```

Comment:
This reads a value from the vector without moving it and, as far as I can tell, the next `push()` will cause it to drop, even though semantically it was moved out of (and therefore should not be dropped).

Reply:
I have modified the pop type T so it only supports ones that implement Copy.